IPInt arguments initialization exceeds available metadata.

MenloDorian opened the pull request:
#### d61ab0d2aa4f32efa6a1bb769515fb856dc80105
<pre>
IPInt arguments initialization exceeds available metadata.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293394">https://bugs.webkit.org/show_bug.cgi?id=293394</a>
<a href="https://rdar.apple.com/149663405">rdar://149663405</a>

Reviewed by Daniel Liu.

Currently, `ipintEntry()` sets up its arguments initialization loop to iterate the total amount of stack
slots reserved and initialize them.  However, the stack slots consists of a number of locals (specified
by IPIntCallee::m_localSizeToAlloc) and a number of rethrow slots (specified by
IPIntCallee::m_numRethrowSlotsToAlloc).  We only have arguments metadata for the locals part, not the
rethrow slots.  As a result, this initialization loop can iterate past the end of the arguments metadata.

This patch will fix this by bounding the iteration at the end of the locals section instead of the end
of the total stack slots.

Covered by existing test JSTests/stress/catch-from-function-inlining-simd-function.js.

* Source/JavaScriptCore/llint/InPlaceInterpreter32_64.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:

Canonical link: <a href="https://commits.webkit.org/295268@main">https://commits.webkit.org/295268@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42d6aff9875b5ad32227f70942a270e7011604ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109748 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55210 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24633 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79368 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19154 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94360 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59694 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18973 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12434 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54581 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97218 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112136 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103154 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31700 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23341 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88411 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32064 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90594 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88081 "Found 1 new API test failure: /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32980 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10762 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16970 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31627 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36967 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31419 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34758 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32979 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

danlliu reviewed (approved):
r=me

ddkilzer commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -188,7 +188,8 @@ const argumINTEnd = csr3
 macro ipintEntry()
-    checkStackOverflow(ws0, argumINTEnd)
+    const argumINTEndAsScratch = argumINTEnd
+    checkStackOverflow(ws0, argumINTEndAsScratch)
```
Why isn’t this same fix needed in the other file?

MenloDorian replied:
This part isn't a fix, but just a way of documenting that we're re-using the `argumINTEnd` register as a scratch here, and not doing anything with the `argumINTEnd` value as the name previously suggested.  We can do without this change, but I thought it added value here.  So, I only added it here.  The 32-bit file was already using a `argumINTTmp` register, and the name implies it's a scratch.  We can't use `argumINTTmp` here because on 64-bit, it aliases with another register which is still live up to that point.

Commit:
IPInt arguments initialization exceeds available metadata.
https://bugs.webkit.org/show_bug.cgi?id=293394
rdar://149663405

Reviewed by Daniel Liu.

Currently, `ipintEntry()` sets up its arguments initialization loop to iterate the total amount of stack
slots reserved and initialize them.  However, the stack slots consists of a number of locals (specified
by IPIntCallee::m_localSizeToAlloc) and a number of rethrow slots (specified by
IPIntCallee::m_numRethrowSlotsToAlloc).  We only have arguments metadata for the locals part, not the
rethrow slots.  As a result, this initialization loop can iterate past the end of the arguments metadata.

This patch will fix this by bounding the iteration at the end of the locals section instead of the end
of the total stack slots.

Covered by existing test JSTests/stress/catch-from-function-inlining-simd-function.js.

* Source/JavaScriptCore/llint/InPlaceInterpreter32_64.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:

Canonical link: https://commits.webkit.org/295268@main
